text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+
+To uniquely identify an `Endpoint`, this crate proposes the `EndpointId` type, which can be implemented as a wrapper around an `Address`.
+
+```rust
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+struct EndpointId(Address);
+```
+
+Note, that this `EndpointId` should be `Clone` and `Copy`, and requires to implement `Eq`, `PartialEq`, and `Hash`, because it is used as a hashmap key in several instances.
+
+## `Command`
+
+`Command`s are messages that are supposed to be issued by higher level business logic like the protocol layer. They are implemented as an `enum`, which in Rust is one way of expressing a polymorphic type:
+
+```rust
+enum Command {
+    // Adds an endpoint to the list.
+    AddEndpoint { url: Url, ... },
+
+    // Removes an endpoint from the list.
+    RemoveEndpoint { epid: EndpointId, ... },
+
+    // Attempts to connect to an endpoint.
+    Connect { epid: EndpointId, ... },
+
+    // Disconnects from an endpoint.
+    Disconnect { epid: EndpointId, ... },
+
+    // Sends a message to a particular endpoint.
+    SendMessage { epid: EndpointId, bytes: Vec<u8>, ... },
+
+    // Sends a message to multiple endpoints.
+    MulticastMessage { epids: Vec<EndpointId>, bytes: Vec<u8>, ...},
+
+    // Sends a message to all endpoints in its list.
+    BroadcastMessage { bytes: Vec<u8>, ... },
+}
+```
+
+This `enum` makes the things the consumer can do with the crate very explicit and descriptive, and also easily extensible.
+
+### `Event`
+
+Similarly to `Command`s, the different kinds of `Event`s in the system are implemented as an `enum` allowing various types of concrete events being sent over event channels:
+
+```rust
+enum Event {
+    // An endpoint was added to the list.
+    EndpointAdded { epid: EndpointId, ... },
+
+    // An endpoint was removed from the list.
+    EndpointRemoved { epid: EndpointId, ... },
+
+    // A new connection is being set up.
+    NewConnection { ep: Endpoint, ... },
+
+    // A broken connection.
+    LostConnection { epid: EndpointId },
+
+    // An endpoint was successfully connected.
+    EndpointConnected { epid: EndpointId, address: Address, ... },
+
+    // An endpoint was disconnected.
+    EndpointDisconnected { epid: EndpointId, ... },
+
+    // A message was sent.
+    MessageSent { epid: EndpointId, num_bytes: usize },
+
+    // A message was received.
+    MessageReceived { epid: EndpointId, bytes: Vec<u8> },
+
+    // A connection attempt should occur.
+    TryConnect { epid: EndpointId, ... },
+}
+
+```
+
+In contrast to commands though, events are messages created by the system, and those that are considered relevant are published for the consumer to execute custom logic.
+
+## Workers
+
+There are two essential asynchronous workers in the system that are running for the node's whole lifetime. That is the `EndpointWorker` and the `TcpWorker`.
```

Comment:
I'd replace `node` with `application` or whatever.

Reply:
Agree. Done.

Reply:
Seems like it's still there :)

Reply:
no it's not (see rendered version). Somehow Gh is not able to track my changes in my `bee-network` branch anymore.

Reply:
Well when I reviewed, the commit that changed it wasn't shown yet... First syndromes of bloated PR :)